CORE-1150 Add Notification message navigation links

psarando opened the pull request:
This PR will refactor the Notification listing so that analysis, app, and data notification messages are now links that navigate the user to the appropriate page.

Other messages are no longer clickable and are not displayed as a link.

Other refactors are also included in this PR, such as standardizing the `NotificationCategory` model and migrating the `getDisplayMessage` function into a `notifications/utils` namespace.

psarando commented on src/components/utils/NameLink.js:
```diff
@@ -38,7 +38,7 @@ const NameLink = React.forwardRef((props, ref) => {
             }}
         >
-            <Highlighter search={searchTerm}>{name}</Highlighter>
+            <Highlighter search={searchTerm || ""}>{name}</Highlighter>
```
Without this change a warning was issued if the `searchTerm` prop was not set in `NameLink`, although `Highlighter` appears to display its child components just fine (like a no-op) with an empty `search` prop.

sri2k1us commented on src/components/data/utils.js:
```diff
@@ -29,7 +29,7 @@ function getEncodedPath(path) {
 const getFolderPage = (folderPath) =>
-    `${constants.PATH_SEPARATOR}${NavigationConstants.DATA}${constants.PATH_SEPARATOR}ds${folderPath}`;
+    `${constants.PATH_SEPARATOR}${NavigationConstants.DATA}${constants.PATH_SEPARATOR}${constants.DATA_STORE_STORAGE_ID}${folderPath}`;
 
```
👍

sri2k1us commented on src/components/layout/Notifications.js:
```diff
@@ -114,7 +112,7 @@ function Notifications(props) {
             let analysisStatus =
                 category.toLowerCase() ===
-                    constants.NOTIFICATION_CATEGORY.ANALYSIS.toLowerCase() ||
+                    NotificationCategory.ANALYSIS.toLowerCase() ||
```
👍

sri2k1us commented on src/components/models/notificationCategory.js:
```diff
@@ -1,12 +0,0 @@
-    team: "Team",
-    community: "Community",
-};
-export default notificationCategory;
```
I thought we will are using file names starting with uppercase only for Components 🤔 . Regardless standardizing this will be good.

ash3rz replied:
I don't think there's any standard or convention really out there, but for some reason my mind seems to think stuff in the models directory seem better suited as PascalCase. I guess because they're supposed to represent some aspect of a data object model, and object = PascalCase to me?  

Our `index` files are also components but they're not PascalCase so 😄 🤷‍♀️ (not saying they should be PascalCase, just an interesting exception)

sri2k1us replied:
Agreed. Having models starting with uppercase seems reasonable to me. WRT `index` files, everyone seems to use lowercase for that. I guess that convention is beyond `react` comes from the `js` community but 🤷‍♂️

psarando replied:
I guess I was also thinking JSON model = object = PascalCase.

I think the `index.js` naming convention goes back to traditional webpage directories, where a URL like `my.site.example.com/some/thing/` would retrieve the document at `my.site.example.com/some/thing/index.html` by default, so an `import SomeThing from "some/thing";` will look for the file `some/thing/index.js` by default. And URLs were traditionally case-insensitive, so I'm used to seeing all lowercase HTML docs and I can't recall ever seeing an `Index.html` file. The `index.js` naming convention is probably also following the same tradition.

So if everyone agrees, I like standardizing on components and models using PascalCase file names.
And things like `util` files, which are just a collection of functions, seem better suited as all lowercase file names.

But I also have no objection to standardizing on lowercase model file names either, and reserving PascalCase strictly for components, if anyone has a strong opinion on that.

sri2k1us replied:
Lets settle for PascalCase for models.

sri2k1us commented on src/components/notifications/Message.js:
```diff
@@ -0,0 +1,137 @@
+
+    return (
+        <Link href={href} as={as} passHref>
+            <NameLink name={message} />
```
Now that we are using `NameLink` for other things as well should this be renamed to something link `DELink` ?

sri2k1us commented on src/components/notifications/Message.js:
```diff
@@ -0,0 +1,137 @@
+
+    return message;
+}
+
```
When I tried this locally, clicking on an `App` Sharing notification caused an `400` error. The address in URL bar was 
`http://localhost:3000/apps?selectedOrder=null&selectedOrderBy=null&selectedPage=null&selectedRowsPerPage=null&selectedFilter=null&selectedCategory={%22system_id%22:%22de%22,%22id%22:%22eeeeeeee-eeee-eeee-eeee-eeeeeeeeeeee%22}`
![image](https://user-images.githubusercontent.com/1250790/97460666-58cf8480-1913-11eb-9736-c0fe1657a71d.png)



I think Clicking on an app notification should take users to the individual app listing using this following `href` and `as`

const href = `/${NavigationConstants.APPS}/[systemId]/[appId]`;	
    const as = `/${NavigationConstants.APPS}/${selectedApp?.system_id}/${selectedApp?.id}`;

psarando replied:
So I was attempting to use `components/apps/utils/getListingPath` by passing `null` for all params except the `category` param, hoping that the code that built or parsed the URL would strip params with `null` values, but it looks like this function is currently building a raw string for the URL and these nulls become the string "null" when parsed by the target page, and I didn't test this in dev mode before opening this PR 😞 

I agree that this link should take the user to the specific app in the listing, so I've made the suggested change, but first I was curious and tried refactoring `getListingPath` to return a URL `{pathname, query}` object compatible with `next/link` and `Router.push`, but it appears `next/link` also preserves `null` values set in that URL's `query` object 😞

sri2k1us reviewed (changes requested):
LGTM but I may have found a bug which needs to investigated / fixed.

ash3rz reviewed (commented):
I didn't catch anything, only had some minor comments 👍

psarando commented on src/components/notifications/Toolbar.js:
```diff
@@ -58,21 +58,21 @@ const NotificationToolbar = (props) => {
+                    "TOOL_REQUEST",
+                    "APPS",
+                    "PERMANENT_ID_REQUEST",
+                    "TEAM",
```
Ah, I had considered it, but that would change the order of the menu items, and I thought they were ordered this way for a specific reason (with `New` on top, followed by `All`, etc.).

ash3rz replied:
🤔 I'd have to test it, but I don't think `Object.keys()` reorders the list of keys. So if you have the `NotificationCategory` object set up with the keys in the order you want to view them in as MenuItems, I think that should still work?

ash3rz replied:
(if not, no big deal either)

psarando replied:
Yes, it does appear to preserve the order of the keys in this model object anyway, but I think it was setup with `All` first, followed by `New`.

I'll go with that and then we can reorder the keys of the category model as needed.

psarando commented:
Thanks for the reviews! 👍 
I think I've addressed all comments and made the suggested changes, so I'll go ahead and merge this PR now, but additional reviews still welcome 🙇
